File: 2019/haskell/Day10.hs
module Main where

import Data.List (mapAccumL, sortOn)
import Data.Maybe (catMaybes, listToMaybe)
import qualified Data.Set as S

data Asteriods = Ast {mx,my :: Int, asteroids :: S.Set (Int,Int)}

inField :: Asteriods -> (Int,Int) -> Bool
inField fld (x,y) = 0 <= x && 0 <= y && x <= mx fld && y <= my fld

-- ordered by l^infinity distance to first lattice point on the line
sightlines :: [(Int,Int)]
sightlines = (0,1):(0,-1):(1,0):(-1,0): rest
  where rest = concatMap f $ (1,1) : concat [concat [[(x,y),(y,x)] | y<- filter (coprime x) [1..x-1]] | x<-[1..]]
        f (x,y) = [(x,y),(x,-y),(-x,y),(-x,-y)]
        coprime x y = gcd x y == 1

(.+) :: (Int,Int) -> (Int,Int) -> (Int,Int)
(a,b) .+ (c,d) = (a+c,b+d)
infixl 6 .+

(.*) :: Int -> (Int,Int) -> (Int,Int)
k .* (a,b) = (k*a,k*b)
infix 7 .*

look :: Asteriods -> (Int,Int) -> (Int,Int) -> Maybe (Int,Int)
look asts ast l = listToMaybe $ filter (flip S.member (asteroids asts)) $ takeWhile (inField asts) $ map (\k -> ast.+k.*l) [1..]

{- part b: we now want to scan clockwise, so sort sightlines as (offsets)
(0,-1),(1,-inf..0),(2,-inf..0)...(1,0),
We do this by calculating their angle in (-pi,pi] from "up"
-}
angle :: (Int,Int) -> Double
{- NB: positive input direction is right and down, we want to order as "up" before "right"
before "down" before "left".
atan2 y x  assumes positive is right and up, and gives clockwise angle from x->+inf, in range (-pi,pi]
atan2 (-x) y, not atan2 y x to measure clockwise from "down", with the branch cut "upwards"
-}
angle (x, y) = atan2 (negate $ fromIntegral x) (fromIntegral y)

vaporise :: Asteriods -> (Int,Int) -> Asteriods
vaporise asts p = asts{asteroids = S.delete p $ asteroids asts}

main :: IO ()
main = do dat' <- lines <$> readFile "../data/day10"
          let dat = concatMap (\(y,l) -> zipWith (\x c -> ((x,y),c)) [0..] l) $ zip [0..] dat'
          let (maxx,maxy) = fst $ last dat
          let asts' = map fst $ filter (\(_,c) -> c =='#') dat
          let asts = Ast maxx maxy $ S.fromList asts'
          let sightlinesBounded = takeWhile (\(lx,ly) -> max (abs lx) (abs ly) <= max maxx maxy) sightlines
          let visibilities = flip map asts' $ \ast -> map (look asts ast) sightlinesBounded
          let (mostVisCount,mostVis) = maximum $ zip (map (length . catMaybes) visibilities) asts'

          putStr "part a: "
          print mostVisCount

          let sightClock = cycle $ sortOn angle sightlinesBounded
          let vaporised = snd $ mapAccumL (\astsLeft l -> maybe (asts,Nothing) (\v -> (vaporise astsLeft v,Just v)) $ look asts mostVis l) asts sightClock
          let betsOn = 200
          let luckyAsteroid = catMaybes vaporised !! (betsOn - 1)
          putStr "part b: "
          print $ 100*fst luckyAsteroid+snd luckyAsteroid

File: 2019/rust/src/bin/day10.rs
use std::cmp::max;
use std::collections::HashSet;
use std::convert::TryInto;
use std::fs;

use num::integer::gcd;

// in increasing distance of first point, bounded by max(w,h)
fn sightlines(w: i32, h: i32) -> Vec<(i32, i32)> {
    let m: i32 = max(w, h);
    let mut res = if m >= 1 {
        vec![
            (1, 0),
            (-1, 0),
            (0, 1),
            (0, -1),
            (1, 1),
            (1, -1),
            (-1, -1),
            (-1, 1),
        ]
    } else {
        Vec::new()
    };
    for x in 1..=m {
        for y in 1..x {
            if gcd(x, y) == 1 {
                res.append(&mut vec![
                    (x, y),
                    (x, -y),
                    (-x, -y),
                    (-x, y),
                    (y, x),
                    (y, -x),
                    (-y, -x),
                    (-y, x),
                ]);
            }
        }
    }
    res
}

// part b: we now want to scan clockwise, so sort sightlines as (offsets)
// (0,-1),(1,-inf..0),(2,-inf..0)...(1,0),
fn sightlines_clockwise(w: i32, h: i32) -> Vec<(i32, i32)> {
    let mut sl = sightlines(w, h);
    let cmp = |&(x, y): &(i32, i32), &(u, v): &(i32, i32)| {
        // NB: positive input direction is right and down, we want to order as "up" before "right"
        // before "down" before "left".
        // atan2(y,x) assumes positive is right and up, and gives clockwise angle from x->+inf, in
        // range (-pi,pi]
        // atan2(-x,y), not atan2(y,x) (written y.atan2(x)) to measure clockwise from "down", with the branch cut "upwards"
        let xf = x as f64;
        let yf = y as f64;
        let uf = u as f64;
        let vf = v as f64;
        ((-xf).atan2(yf)).partial_cmp(&(-uf).atan2(vf)).unwrap()
    };
    sl.sort_unstable_by(cmp);
    sl
}

fn main() {
    let field = fs::read_to_string("../data/day10").unwrap();
    let h: i32 = field.lines().next().unwrap().len().try_into().unwrap();
    let w: i32 = field.lines().count().try_into().unwrap();
    let sightlines = sightlines(w, h);
    let furthest_edge = |&(x, y)| max(max(x, w - 1 - x), max(y, h - 1 - y));
    let asts: HashSet<(i32, i32)> = field
        .lines()
        .enumerate()
        .flat_map(|(y, l)| {
            l.chars()
                .enumerate()
                .filter(|(_, c)| *c == '#')
                .map(move |(x, _)| (x.try_into().unwrap(), y.try_into().unwrap()))
        })
        .collect();
    let get_sight = |asts: &HashSet<_>, ast: &(i32, i32), (sx, sy)| -> Option<(i32, i32)> {
        for k in 1.. {
            let asxk = ast.0 + k * sx;
            let asyk = ast.1 + k * sy;
            if asxk > w || asyk > h || asxk < 0 || asyk < 0 {
                return None;
            };
            if asts.contains(&(asxk, asyk)) {
                return Some((asxk, asyk));
            }
        }
        return None;
    };

    let sees = asts.iter().map(|ast| {
        let m: i32 = furthest_edge(ast).try_into().unwrap();
        let mut see = 0;
        for (sx, sy) in sightlines.iter().take_while(|(sx, sy)| max(sx, sy) <= &m) {
            match get_sight(&asts, ast, (sx, sy)) {
                Some(_) => see += 1,
                None => {}
            }
        }
        (see, ast)
    });
    let (best_sees, best_ast) = sees.max().unwrap();
    let best_ast = &best_ast.clone();
    println!("part a: {:?} sees {} other asteroids", best_ast, best_sees);

    let bets_on = 200;
    let vaporise_lines = sightlines_clockwise(w, h);
    let mut vap = 0;
    let mut remaining = asts;
    for (lx, ly) in vaporise_lines.iter().cycle() {
        match get_sight(&remaining, best_ast, (lx, ly)) {
            Some(a) => {
                vap += 1;
                remaining.remove(&a);
                if vap == bets_on {
                    println!("part b: bet resolution, the {}th asteroid to be vaporised was {:?}, for an answer of {}",bets_on,a,100*a.0+a.1);
                    break;
                }
            }
            None => {}
        }
    }
}
